lib/request_started_on_middleware.rb

```diff
@@ -0,0 +1,59 @@
+# Add to config/application.rb:
+#
+#     config.middleware.use 'RequestStartedOnMiddleware'
+#
+class RequestStartedOnMiddleware
+  def initialize(app)
+    @app = app
+  end
+
+  def call(env)
+    begin
+      start_request(env['PATH_INFO'], Time.now.utc)
+      response = @app.call(env)
+    ensure
+      complete_request
```

Comment:
Seems like you could avoid a lot of complexity with this by just logging as part of the middleware.

```ruby
def call(env)
  start_time, end_time = [Time.now, nil]
  begin
    @app.call(env)
  ensure
    log_to_request_time_dot_log(start_time, end_time) if too_long?
  end
end
```

The extra `Thread` seems like a much larger over complication of what you are trying to do here, and I am unsure why it is needed in the first place.

Reply:
Agreed, the thread part is the complicated part but is not without good reason.  The problem is that the thread is hung and sometimes doesn't ever finish because it's deadlocked, killed before finishing, or finishes much later than the user gets the 502 error.  In all three of these cases, we won't get to the `log_to_request_time_dot_log` method until after the user has captured or investigate the logs.

Reply:
> doesn't ever finish because it's deadlocked, killed before finishing, or finishes much later

Fair enough.  I have some points below, but this is mostly just followup, and my real status on the review is [here](https://github.com/ManageIQ/manageiq/pull/17842#issuecomment-412658508):

* * *

So for these three:

* "deadlocked":  valid
* "killed before finishing":  true
* "finishes much later than the user gets the 502":  Well, this would still get logged, the data from the request would just end up being sent into the void.

That said, with the first two (which the first, the _deadlock_ would probably lead to the second happening eventually), I want to say that I have seen when I have killed a process (at least with a `Ctrl-C`), that the "Completed" message still ends up in the logs.  I have to do some digging into the Rails source a bit more to see how `ActiveSupport::LogSubscriber` handles control signals and allow this to happen, since that seems to be what is handling this here.

Curious if some form a `rescue`/`at_exit` would allow this to work, but since you are trying to be proactive long term, this might just be more of a curiosity/musing for myself than anything.

Reply:
While I agree, the "eventually" logs `Completed` is something we can parse logs for after the fact... I've seen too many logs where they never `Completed` due to deadlock, killed before finishing or the user captured the logs before it was logged.

The ONLY reason to do this "clever" in memory thread local variable thing is because we don't get the log message and therefore can't trust that we'll get to the rack request completion point in the thread and thus we need to use another thread to track long requests.  If there is another way, I'm open to suggestions.